feat : Task 저장 기능

yellTa opened the pull request:
task객체를 생성하고 저장하는 로직

## 흐름
1. "/task" URI를 통해 Post로 데이터 전달받기
2. Servlet의 request에서 데이터 받기[
3. 받은 데이터를 TaskDao를 통해 DB에 저장하기

Commit:
feat : Task객체생성

데이터를 저장할 때 사용하는 객체의 생성자는
title, date, owner, priority만 생성하고 저장
-> userId는 auto_increment값

데이터를 조회할 때 사용할 객체의 생성자는 json으로 파싱하기 위해
오버로딩을 사용해 user_id 추가

Commit:
Task 객체 수정
Task의 date를 String이 아닌 Date객체로 변경

Commit:
feat body의 json 데이터를 Task객체로 변환하는 로직 TaskServlet에 추가

askDao
데이터를 저장하기 위한 메서드 추가

Task.java
객체를 json으로 변환할 때 날짜, 시간을 특정한 형식으로 표현하기 위해 @JsonFormat사용

TaskServlet.java
request를 통해 body를 읽은 후
jsonString을 Task객체로 변환
변환 과정 중 Date속성은 Task객체에 @JsonFomat을 사용하여 String이 아닌 Date객체로 변환

이를 위해 pom.xml에 jackson.datatype 라이브러리 의존성 추가

pom.xml
jackson.datatype 라이브러리 추가

Commit:
feat: TaskDao save로직 구현하기

JDBC를 이용해 저장
Task에서 날짜 + 시간 객체가 아닌 날짜 데이터로 data 컬럼 저장
Task.java java Date가 아닌 java.sql.Date로 객체 지정

Commit:
docs API명세서 수정, feat: 사용하지 않는 생성자 제거

docs: api 명세서 수정
task에서 post로 값을 보낼떄 Progress를 사용하지 않고 보내는 것으로 수정

feat : 사용하지 않는 생성자 제거
Task.java에서 사용하지 않는 생성자 제거

cohajun commented on .idea/uiDesigner.xml:
@yellTa 
이 파일은 리뷰에 필요한걸까요?
gitignore 파일에 대해 알아보고 적용해보세요~!

cohajun commented on src/main/java/boost/chaper2/dao/TaskDao.java:
```diff
@@ -0,0 +1,34 @@
+    public void saveData(Task task) {
+
+        try{
+            Class.forName("com.mysql.cj.jdbc.Driver");
```
@yellTa 
saveData를 호출할 때마다 
Class.forName("com.mysql.cj.jdbc.Driver");
을 계속 로드하는게 맞을까요~?

yellTa replied:
Class.forName("com.mysql.cj.jdbc.Driver") 부분은 항상 재사용되는 부분이니까 싱글톤 패턴으로 구현한 뒤 재사용하는 방법으로 사용하면 될 것 같습니다! 해당 부분 조금 더 조사해본 뒤 수정해보곘습니다!

yellTa replied:
static block 에 넣어서 connectino을 생성하는 것으로 변경했습니다! 하지만 궁금한 점이 생겼는데요!

connection을 연결한 후 메모리 누수 방지를 위해서 connection은 닫아야 하는 것으로 알고있습니다.

TaskDao에서 로직을 수행하고 닫아주기
애플리케이션이 종료될 때 닫아주기

위 두 가지 방법중 저는 아래를 택했습니다. TaskDao가 수행될 떄 마다 connection을 닫아주면 static 블록을 사용하는 의미가 없기 때문이라고 생각해서요

위 상황에서는 어떤식으로 처리해야할까요?

cohajun replied:
@yellTa 
```
public class DatabaseConnector {
	private static final String DB_URL = "";
	private static final String DB_USER = "";
	private static final String DB_PASSWORD = "";

	static {
		try {
			Class.forName("com.mysql.jdbc.Driver");
		} catch (ClassNotFoundException c) {
			throw new RuntimeException(c);
		}
	}

	public static Connection getConnection() throws SQLException {
		return DriverManager.getConnection(DB_URL, DB_USER, DB_PASSWORD);
	}
}
```
컴파일은 안해봤지만 이런식으로 작성할 것 같습니다~
드라이버는 디비에 연결할 때 한번만 호출될거에요~ 하지만 커넥션은 그렇지 않습니다!
현재는 dbcp를 사용하지 않지만 dbcp에 대해서도 공부해보면 좋을 것 같아요~

cohajun commented on src/main/java/boost/chaper2/dao/TaskDao.java:
```diff
@@ -0,0 +1,34 @@
+
+        try{
+            Class.forName("com.mysql.cj.jdbc.Driver");
+        }catch(Exception e){
+            e.printStackTrace();
+        }
```
@yellTa 
아래 catch문과 indent가 안맞는것으로 보이는데요~
indent를 통일할 수 있는 방법을 찾아보세요!
(힌트 codestyle)

yellTa replied:
indent 맞춰서 수정해보곘습니다!

cohajun commented on src/main/java/boost/chaper2/dao/TaskDao.java:
```diff
@@ -0,0 +1,34 @@
+        }catch(Exception e){
+            e.printStackTrace();
+        }
+        String sql = "Insert into tasks (title, date, owner, priority) values(?,?,?,?)";
```
@yellTa 
이건 좀 의아한데요~
보통 소문자로 예약어(insert into table )를 작성하거나
대문자로 예약어(INSERT INTO table )를 작성하는데
Camel case가 작동을 하나요? 
작동하더라도 일반적이진 않은 것 같아요~
통일해주시면 좋을 것 같습니다.

yellTa replied:
Camel case로도 작동을 합니다... 일반적인 규약인데 막상 작성하면서 크게 신경쓰지 못한 부분 같습니다 대문자로 통일해서 수정해보겠습니다!

cohajun commented on src/main/java/boost/chaper2/dto/Task.java:
```diff
@@ -0,0 +1,48 @@
+    int user_id;
+    String title;
+    @JsonFormat(pattern = "yyyy.MM.dd")
+    java.sql.Date date;
```
@yellTa 
java.sql.Date를 쓰려는게 맞을까요~?
java.util.Date를 쓰려고 의도했던 것 같은데 한번 확인해주세요~
인텔리제이에 auto import 기능을 설정해주세요

yellTa replied:
해당 부분은 PreparedStatement에서 setDate는 java.sql.Date 타입을 요구해서 처음에 받을 때 java.sql.Date타입으로 받았습니다!

혹시 자바의 java.util.Date로 받고 DAO부분에서 sql.Date로 변환 후 setDate()로 저장하는 편이 더 나을까요??

cohajun replied:
@yellTa 
질문 주신 내용에 있어서 java.util.Date 와 java.sql.Date 의 차이가 어떤 부분이 있는지 찾아보고 제가 리뷰 드린 내용을 모두 반영할 필요는 없으니 어떤 클래스를 선택하는지에 있어서 선택하신 이유를 작성해주시면 좋을 것 같습니다~!

yellTa replied:
commit 422b7e3b81c67fc4e08f58bc886e75391d3467cb
해당 내용 커밋했습니다!

DB와 데이터를 매핑할때 java.util.Date가 아니라 java.sql.Date를 사용해야 했습니다. 따라서 처음부터 받을 때 java.sql.Date를 사용해서 객체를 생성하고 Servlet에서도 그런 방법으로 받았습니다.

하지만 만약 애플리케이션 내에서 날짜 데이터를 가져와 가공해야하는 상황이 벌어진다면 java.util.Date를 사용하는게 맞다고 판단해 TaskDto와 TaskServlet에서는 java.util.Date를 사용했습니다.

TaskDao에서는 데이터베이스에 읽기 쓰기 작업을 수행해야 하기 때문에 java.util.Date로 받은 객체를 java.sql.Date로 변환해 데이터 베이스의 데이터 타입과 매핑시켰습니다!

ps: java.sql.Date는 java.util.Date의 서브클래스라고 합니다

cohajun commented on src/main/java/boost/chaper2/dto/Task.java:
```diff
@@ -0,0 +1,48 @@
+        this.progress = progress;
+    }
+
+    public int getUser_id() {
+        return user_id;
+    }
+
+    public String getTitle() {
+        return title;
+    }
+
+    public java.sql.Date getDate() {return date;}
+
+    public String getOwner() {
+        return owner;
+    }
+
+    public int getPriority() {
+        return priority;
+    }
```
@yellTa 
getter만 제공하려고 했던 의도는 좋네요 👍
근데 필드의 접근제어자가 default네요~?
의도된걸까요?

yellTa replied:
필드의 접근 제어자는 외부 패키지에서 수정할 수 없도록 private이 되어야합니다! 깊게 생각하지 못한 부분이었습니다 해당 부분 private로 수정하겠습니다!

cohajun commented on src/main/java/boost/chaper2/servlet/TaskServlet.java:
```diff
@@ -0,0 +1,47 @@
+        objectMapper.disable(SerializationFeature.WRITE_DATES_AS_TIMESTAMPS);
+        Task task = objectMapper.readValue(jsonString, Task.class);
+
+        TaskDao dao = new TaskDao();
```
@yellTa 
TaskDao는 thread safe 할까요?

cohajun replied:
@yellTa 
저한테는 댓글이 안보이는데요~
다시 달아주시겠어요?!

Commit:
refact : 리뷰 받은 사항 업데이트

1. git ignore
ide 설정 파일 경로를 .gitignore 에 추가

2. 드라이버 중복 호출
DatabaseManager.java를 생성해 싱글톤으로 객체 생성 하나의 인스턴스만 생성한다.

3. indent 수정
TaskDao.java에 있던 try catch문 indent 변경

4. SQL 예약어 사용
TaskDao.java에서 사용하던 SQL구문 예약어를 사용하도록 변경
INSERT INTO ...

5. java.sql.Date 와 java.util.Date사용방안
PreparedStatement에서 요구하는 파라미터 형식은 java.sql.Date
java.util.Date를 사용하려면 Dao에서 객체 변환후 사용해야 함

6. Task의 접근제어자
외부에서 변경할 수 없도록 private로 변경

7. thread safe부분은 아직 수정중

Commit:
refact : static 블록을 사용한 Driver Connection

TaskDao.java 에서 static블록을 사용해 Driver Connection을 얻음
Connection을 닫기위해 closeConnection메서드추가
cleanUpSQLDriver 메소드를 사용해서 JVM에 등록된 모든 JDBC드라이버를 제거

AppListner.java
애플리케이션이 종료될 떄 TaskDao.java의 작성한 closeConnection을 이용해 connection 종료
 cleanUpSQLDriver를 이용해 JVM에 등록된 모든 JDBC드라이버 해지

Commit:
refact: static 블록을 사용해 Connection객체 생성

static블록을 사용해 JDBC드라이버 로드
클래스가 로드 될 떄 처음 생성되고 재사용함
try catch resource 구문을 사용해 자원을 닫아줌

Commit:
refact: Task, TaskServlet에서 java.util.Date사용 Dao에서 java.sql.Date사용

Servlet과 DTO에서는 날짜 데이터를 가공할 확률이 있어 util.Date사용

java.sql.Date는 Dao에서 데이터베이스의 DATE와 매핑시키기위해 java.util.Date를 sql.Date로 변환해 사용

Commit:
style: TaskServlet- 사용하지 않는 라이브러리 import 제거, Task- indent 수정

TaskServlet.java에서 사용하지 않는 라이브러리 제거
1. jackson.databind.SerializationFeature
JSON 변환 과정중 날짜 및 시간과 같은 특수한 데이터 처리
2. jackson.datatype.jsr310.JavaTimeModule
날짜 및 시간 API처리 서포트 모듈

Task.java 들여쓰기 수정

cohajun commented on .gitignore:
```diff
@@ -8,6 +8,7 @@ target/
 .idea/jarRepositories.xml
 .idea/compiler.xml
 .idea/libraries/
+.idea
```
@yellTa 
.idea/ 
로 적용하면 해당 폴더 모두 gitignore 처리될 것 같아요~ 테스트해보시고 적용해주세요

Commit:
docs: 필요없는 파일 삭제

Commit:
docs : gitmessage파일 추가
